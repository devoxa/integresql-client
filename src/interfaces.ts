export interface IntegreSQLDatabaseConfig {
  username: string
  password: string
  host: string
  port: number
  database: string
}

export interface InitializeTemplateResponse {
  database: {
    templateHash: string
    config: IntegreSQLDatabaseConfig
  }
}

export interface GetTestDatabaseResponse {
  id: number
  database: {
    templateHash: string
    config: IntegreSQLDatabaseConfig
  }
}
