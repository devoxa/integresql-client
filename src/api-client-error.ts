export class IntegreSQLApiClientError extends Error {
  responseStatus: number
  responseText: string
}

export interface CreateIntegreSQLApiClientErrorOptions {
  message: string
  responseStatus: number
  responseText: string
}

export function createIntegreSQLApiClientError(options: CreateIntegreSQLApiClientErrorOptions) {
  const error = new IntegreSQLApiClientError(options.message)
  error.responseStatus = options.responseStatus
  error.responseText = options.responseText
  return error
}
