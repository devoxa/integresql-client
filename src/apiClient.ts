import { createIntegreSQLApiClientError } from './apiClientError'
import { GetTestDatabaseResponse, InitializeTemplateResponse } from './interfaces'

export interface IntegreSQLApiClientOptions {
  url: string
}

export class IntegreSQLApiClient {
  baseUrl: string

  constructor(options: IntegreSQLApiClientOptions) {
    const url = new URL(options.url)
    this.baseUrl = url.toString() + 'api/v1'
  }

  async request<T>(
    method: 'GET' | 'POST' | 'PUT' | 'DELETE',
    endpoint: string,
    body?: Record<string, unknown>
  ): Promise<T> {
    const url = this.baseUrl + endpoint

    const response = await fetch(url, {
      method,
      body: body ? JSON.stringify(body) : undefined,
      headers: { 'Content-Type': 'application/json' },
    })

    const text = await response.text()

    if (response.status >= 400) {
      throw createIntegreSQLApiClientError({ responseStatus: response.status, responseText: text })
    }

    return text ? (JSON.parse(text) as T) : (null as T)
  }

  async initializeTemplate(hash: string): Promise<InitializeTemplateResponse> {
    return this.request<InitializeTemplateResponse>('POST', '/templates', { hash })
  }

  async finalizeTemplate(hash: string): Promise<null> {
    return this.request<null>('PUT', `/templates/${hash}`)
  }

  async discardTemplate(hash: string): Promise<null> {
    return this.request<null>('DELETE', `/templates/${hash}`)
  }

  async discardAllTemplates(): Promise<null> {
    return this.request<null>('DELETE', '/admin/templates')
  }

  async getTestDatabase(hash: string): Promise<GetTestDatabaseResponse> {
    return this.request<GetTestDatabaseResponse>('GET', `/templates/${hash}/tests`)
  }

  async reuseTestDatabase(hash: string, id: number): Promise<null> {
    return this.request<null>('DELETE', `/templates/${hash}/tests/${id}`)
  }

  async recreateTestDatabase(hash: string, id: number): Promise<null> {
    return this.request<null>('POST', `/templates/${hash}/tests/${id}/recreate`)
  }
}
