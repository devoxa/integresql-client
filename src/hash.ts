import crypto from 'crypto'
import glob from 'fast-glob'
import fs from 'fs'
import path from 'path'

export async function sha1HashFiles(globPatterns: Array<string>): Promise<string> {
  const files = await glob(globPatterns)
  const filePaths = files.map((file) => path.join(process.cwd(), file))

  const fileHashes = await Promise.all(filePaths.map(sha1HashFile))
  return sha1HashString(fileHashes.join(''))
}

export function sha1HashString(value: string): string {
  return crypto.createHash('sha1').update(value).digest('hex')
}

export function sha1HashFile(filePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const hash = crypto.createHash('sha1')
    const readStream = fs.createReadStream(filePath)

    // istanbul ignore next
    readStream.on('error', (err) => reject(err))

    readStream.on('data', (chunk) => hash.update(chunk))
    readStream.on('end', () => resolve(hash.digest('hex')))
  })
}
