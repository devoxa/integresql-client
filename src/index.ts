export * from './api-client'
export * from './api-client-error'
export * from './client'
export * from './interfaces'
