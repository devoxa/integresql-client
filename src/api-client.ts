import fetch from 'node-fetch'
import path from 'path'
import { createIntegreSQLApiClientError } from './api-client-error'
import { GetTestDatabaseResponse, InitializeTemplateResponse } from './interfaces'

export interface IntegreSQLApiClientOptions {
  url: string
}

export class IntegreSQLApiClient {
  baseUrl: string

  constructor(options: IntegreSQLApiClientOptions) {
    const url = new URL(options.url)
    this.baseUrl = url.toString() + 'api/v1'
  }

  async request<T>(
    method: 'GET' | 'POST' | 'PUT' | 'DELETE',
    endpoint: string,
    body?: Record<string, unknown>
  ): Promise<T> {
    const url = path.join(this.baseUrl, endpoint)

    const response = await fetch(url, {
      method,
      body: body ? JSON.stringify(body) : undefined,
      headers: { 'Content-Type': 'application/json' },
    })

    const text = await response.text()

    if (response.status >= 400) {
      throw createIntegreSQLApiClientError({ responseStatus: response.status, responseText: text })
    }

    return text ? JSON.parse(text) : null
  }

  async initializeTemplate(hash: string) {
    return this.request<InitializeTemplateResponse>('POST', '/templates', { hash })
  }

  async finalizeTemplate(hash: string) {
    return this.request<null>('PUT', `/templates/${hash}`)
  }

  async discardTemplate(hash: string) {
    return this.request<null>('DELETE', `/templates/${hash}`)
  }

  async discardAllTemplates() {
    return this.request<null>('DELETE', `/admin/templates`)
  }

  async getTestDatabase(hash: string) {
    return this.request<GetTestDatabaseResponse>('GET', `/templates/${hash}/tests`)
  }

  async reuseTestDatabase(hash: string, id: number) {
    return this.request<null>('DELETE', `/templates/${hash}/tests/${id}`)
  }
}
