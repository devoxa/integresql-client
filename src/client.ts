import { IntegreSQLApiClient } from './api-client'
import { IntegreSQLApiClientError } from './api-client-error'
import { sha1HashFiles } from './hash'
import { IntegreSQLDatabaseConfig } from './interfaces'

export interface IntegreSQLClientOptions {
  url: string
}

export class IntegreSQLClient {
  api: IntegreSQLApiClient

  constructor(options: IntegreSQLClientOptions) {
    this.api = new IntegreSQLApiClient({ url: options.url })
  }

  async initializeTemplate(
    hash: string,
    setupCallback: (databaseConfig: IntegreSQLDatabaseConfig) => Promise<void>
  ): Promise<void> {
    // Attempt to create a new template database identified through the hash
    let databaseConfig
    try {
      const response = await this.api.initializeTemplate(hash)
      databaseConfig = response.database.config
    } catch (err) {
      // Some other process has already created a template database for this hash,
      // so we can just consider it ready at this point.
      if (err instanceof IntegreSQLApiClientError && err.responseStatus === 423) {
        return
      }

      throw err
    }

    // The template database is created and ready for setup (apply migrations / seed fixtures)
    try {
      await setupCallback(databaseConfig)
    } catch (err) {
      // If we encountered any errors during setup we discard the template database
      await this.api.discardTemplate(hash)

      throw err
    }

    // The template database is setup and ready for tests
    await this.api.finalizeTemplate(hash)

    // Warm up the template database, since getting the first test database can take a second or two
    await this.getTestDatabase(hash)
  }

  async getTestDatabase(hash: string): Promise<IntegreSQLDatabaseConfig> {
    const response = await this.api.getTestDatabase(hash)
    return response.database.config
  }

  // --- Helpers --- //

  async hashFiles(globPatterns: Array<string>): Promise<string> {
    return sha1HashFiles(globPatterns)
  }

  databaseConfigToConnectionUrl(databaseConfig: IntegreSQLDatabaseConfig): string {
    const { username, password, host, port, database } = databaseConfig
    return `postgresql://${username}:${password}@${host}:${port}/${database}`
  }
}
