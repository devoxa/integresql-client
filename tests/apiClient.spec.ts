import { IntegreSQLApiClient } from '../src/apiClient'
import { IntegreSQLApiClientError } from '../src/apiClientError'
import {
  GetTestDatabaseResponse,
  InitializeTemplateResponse,
  IntegreSQLDatabaseConfig,
} from '../src/interfaces'

const mockFetch = jest.fn()
global.fetch = mockFetch

function getMockFetchImplementation(status: number, response: string) {
  const mockFetchImplementation = async () => ({
    status: status,
    text: async () => response,
  })

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return mockFetchImplementation as any
}

const client = new IntegreSQLApiClient({ url: 'http://localhost:5000' })

const mockDatabaseConfig: IntegreSQLDatabaseConfig = {
  username: 'user',
  password: 'pass',
  host: 'localhost',
  port: 5432,
  database: 'mock_database_0',
}

describe('apiClient', () => {
  beforeEach(() => {
    mockFetch.mockReset()
  })

  test('sets the correct base URL', () => {
    const withTrailingSlash = new IntegreSQLApiClient({ url: 'http://localhost:5000/' })
    const withoutTrailingSlash = new IntegreSQLApiClient({ url: 'http://localhost:5000' })

    expect(withTrailingSlash.baseUrl).toEqual('http://localhost:5000/api/v1')
    expect(withoutTrailingSlash.baseUrl).toEqual('http://localhost:5000/api/v1')
  })

  test('errors when the HTTP response status indicates an issue', async () => {
    const mockResponse = { message: 'template is already initialized' }
    mockFetch.mockImplementation(getMockFetchImplementation(423, JSON.stringify(mockResponse)))

    let error
    try {
      await client.initializeTemplate('mock-hash')
    } catch (err) {
      error = err as IntegreSQLApiClientError
    }

    expect(error).toBeInstanceOf(IntegreSQLApiClientError)
    expect(error?.message).toMatchSnapshot()
    expect(error?.responseStatus).toEqual(423)
    expect(error?.responseText).toEqual(JSON.stringify(mockResponse))
  })

  describe('endpoints', () => {
    test('can initialize a template', async () => {
      const mockResponse: InitializeTemplateResponse = {
        database: { templateHash: 'mock-hash', config: mockDatabaseConfig },
      }
      mockFetch.mockImplementation(getMockFetchImplementation(200, JSON.stringify(mockResponse)))

      const response = await client.initializeTemplate('mock-hash')

      expect(response).toEqual(mockResponse)
      expect(mockFetch.mock.calls).toMatchSnapshot()
    })

    test('can finalize a template', async () => {
      mockFetch.mockImplementation(getMockFetchImplementation(200, ''))

      const response = await client.finalizeTemplate('mock-hash')

      expect(response).toEqual(null)
      expect(mockFetch.mock.calls).toMatchSnapshot()
    })

    test('can discard a template', async () => {
      mockFetch.mockImplementation(getMockFetchImplementation(200, ''))

      const response = await client.discardTemplate('mock-hash')

      expect(response).toEqual(null)
      expect(mockFetch.mock.calls).toMatchSnapshot()
    })

    test('can discard all templates', async () => {
      mockFetch.mockImplementation(getMockFetchImplementation(200, ''))

      const response = await client.discardAllTemplates()

      expect(response).toEqual(null)
      expect(mockFetch.mock.calls).toMatchSnapshot()
    })

    test('can get a test database', async () => {
      const mockResponse: GetTestDatabaseResponse = {
        id: 1,
        database: { templateHash: 'mock-hash', config: mockDatabaseConfig },
      }
      mockFetch.mockImplementation(getMockFetchImplementation(200, JSON.stringify(mockResponse)))

      const response = await client.getTestDatabase('mock-hash')

      expect(response).toEqual(mockResponse)
      expect(mockFetch.mock.calls).toMatchSnapshot()
    })

    test('can reuse a test database', async () => {
      mockFetch.mockImplementation(getMockFetchImplementation(200, ''))

      const response = await client.reuseTestDatabase('mock-hash', 42)

      expect(response).toEqual(null)
      expect(mockFetch.mock.calls).toMatchSnapshot()
    })

    test('can recreate a test database', async () => {
      mockFetch.mockImplementation(getMockFetchImplementation(200, ''))

      const response = await client.recreateTestDatabase('mock-hash', 42)

      expect(response).toEqual(null)
      expect(mockFetch.mock.calls).toMatchSnapshot()
    })
  })
})
