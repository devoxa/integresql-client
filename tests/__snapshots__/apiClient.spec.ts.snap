// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiClient endpoints can discard a template 1`] = `
[
  [
    "http://localhost:5000/api/v1/templates/mock-hash",
    {
      "body": undefined,
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "DELETE",
    },
  ],
]
`;

exports[`apiClient endpoints can discard all templates 1`] = `
[
  [
    "http://localhost:5000/api/v1/admin/templates",
    {
      "body": undefined,
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "DELETE",
    },
  ],
]
`;

exports[`apiClient endpoints can finalize a template 1`] = `
[
  [
    "http://localhost:5000/api/v1/templates/mock-hash",
    {
      "body": undefined,
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PUT",
    },
  ],
]
`;

exports[`apiClient endpoints can get a test database 1`] = `
[
  [
    "http://localhost:5000/api/v1/templates/mock-hash/tests",
    {
      "body": undefined,
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`apiClient endpoints can initialize a template 1`] = `
[
  [
    "http://localhost:5000/api/v1/templates",
    {
      "body": "{"hash":"mock-hash"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`apiClient endpoints can recreate a test database 1`] = `
[
  [
    "http://localhost:5000/api/v1/templates/mock-hash/tests/42/recreate",
    {
      "body": undefined,
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`apiClient endpoints can reuse a test database 1`] = `
[
  [
    "http://localhost:5000/api/v1/templates/mock-hash/tests/42",
    {
      "body": undefined,
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "DELETE",
    },
  ],
]
`;

exports[`apiClient errors when the HTTP response status indicates an issue 1`] = `"API request to IntegreSQL failed: {"message":"template is already initialized"} (Status 423)"`;
